src/Stratis.Bitcoin/Base/ChainHeadersBehavior.cs

```diff
@@ -154,195 +155,260 @@ protected override void DetachCore()
-                                            this.InvalidHeaderReceived = true;
-                                            reorgPrevented = true;
-                                        }
-                                        else this.logger.LogTrace("Reorganization of length {0} accepted, consensus tip is '{1}'.", reorgLength, consensusTip);
-                                    }
-                                }
-                            }
+                    if (prev == null)
+                    {
+                        this.logger.LogTrace("Previous header of the new header '{0}' was not found on our chain either.", header);
 
-                            // Switch to better chain.
-                            if (!reorgPrevented)
-                            {
-                                this.logger.LogTrace("New chain tip '{0}' selected, chain work is '{1}'.", this.pendingTip, this.pendingTip.ChainWork);
-                                this.Chain.SetTip(this.pendingTip);
-                            }
-                        }
+                        // If we can't connect the header we received from the peer, we might be on completely different chain or
+                        // a reorg happened recently. If we ignored it, we would have invalid view of the peer and the propagation
+                        // of blocks would not work well. So we ask the peer for headers using "getheaders" message.
+                        // Enforce a sync.
+                        doTrySync = true;
+                        break;
+                    }
+
+                    // Now we know the previous block header and thus we can connect the new header.
+                }
+
+                tip = new ChainedBlock(header, header.GetHash(peer.Network.NetworkOptions), prev);
+                bool validated = this.Chain.GetBlock(tip.HashBlock) != null || tip.Validate(peer.Network);
+                validated &= !this.chainState.IsMarkedInvalid(tip.HashBlock);
+                if (!validated)
+                {
+                    this.logger.LogTrace("Validation of new header '{0}' failed.", tip);
+                    this.InvalidHeaderReceived = true;
+                    break;
+                }
+
+                this.pendingTip = tip;
+            }
+
+            if (pendingTipBefore != this.pendingTip)
+                this.logger.LogTrace("Pending tip changed to '{0}'.", this.pendingTip);
 
-                        ChainedBlock chainedPendingTip = this.pendingTip == null ? null : this.Chain.GetBlock(this.pendingTip.HashBlock);
-                        if (chainedPendingTip != null)
+            if ((this.pendingTip != null) && (this.pendingTip.ChainWork > this.Chain.Tip.ChainWork))
+            {
+                // Long reorganization protection on POS networks.
```

Comment:
Do you want a trace log here? I guess the only condition you don't have a trace on is if maxReorgLength == 0, is that a condition that might be useful to have in the trace log?

Reply:
probably not really, either reorg protection is enabled (stratis) or disabled (bitcoin) and that's about it

but still, old code :D